### ISSUE TITLE
`org-mode` is broken after restarting emacs on latest develop when emacs < 29

### ISSUE THREAD

**Issue by nixmaniack** [syl20bnr/spacemacs#15896]
#### Description :octocat:
`org-mode` is broken. Opening an org file doesn't load the `org-mode`.

I tried removing all org packages, restart emacs to let it install all the packages and then open an org file which works for that session. If you quit and restart it throws `(invalid-function org-assert-version)` error.

#### Reproduction guide :beetle:
- Delete `org-9.6.1` and `org-contrib-0.4.1` from `~/.emacs.d/elpa/28.2/develop`.
- Start emacs.
- Let it install packages.
- Open any org file, this loads org-mode file for the session.
- Restart emacs.
- Open any org file, it throws error.

*Observed behaviour:* :eyes: :broken_heart:
Opening an org file throws following error.
```
File mode specification error: (invalid-function org-assert-version)
```

*Expected behaviour:* :heart: :smile:
Opens file and applies org-mode and related layer configuration.

#### System Info :computer:
- OS: gnu/linux
- Emacs: 28.2.50
- Spacemacs: 0.999.0
- Spacemacs branch: develop (rev. d08822c61)
- Graphic display: t
- Running in daemon: nil
- Distribution: spacemacs
- Editing style: hybrid
- Completion: ivy
- Layers:
```elisp
(auto-completion emacs-lisp git ivy lsp markdown multiple-cursors org
                 (shell :variables shell-default-height 30 shell-default-position 'bottom shell-default-shell 'vterm)
                 syntax-checking version-control treemacs themes-megapack)
```
- System configuration features: ACL CAIRO DBUS FREETYPE GIF GLIB GMP GNUTLS GPM GSETTINGS HARFBUZZ IMAGEMAGICK JPEG JSON LCMS2 LIBOTF LIBSELINUX LIBSYSTEMD LIBXML2 M17N_FLT MODULES NATIVE_COMP NOTIFY INOTIFY PDUMPER PNG RSVG SECCOMP SOUND THREADS TIFF TOOLKIT_SCROLL_BARS X11 XDBE XIM XPM GTK3 ZLIB

**Comment by lebensterben** [comment]
cc @smile13241324

**Comment by lebensterben** [comment]
Confirmed. At first my emacs was working fine. But after deleting the two packages mentioned above, Emacs won't start properly. (spacemac-buffer won't show up since it requires org with my settings)

**Comment by lebensterben** [comment]
Inspired by https://irreal.org/blog/?p=10999

To fix it:

1. `cd ~/.emacs.d; rm -rf elpa/28.2/develop/org-9*`

2. start emacs with `emacs -Q` and evaluate (with `C-j`)

```elisp
(require 'package)
(add-to-list 'package-archives '("melpa" . "https://melpa.org/packages/"))
(setq package-user-dir
      (expand-file-name "develop"
                        (expand-file-name emacs-version
                                          (expand-file-name "elpa"
                                                            user-emacs-directory))))
(package-refresh-contents)
(package-install 'org)
```

3. Restart Emacs normally.

**Comment by nixmaniack** [comment]
@lebensterben I followed your workaround but looks it didn't work for me. I did `rm -rf` the org packages but the last line to install `org` throws this message.

```
"‘org’ is already installed"
```

I just followed your exact instruction and haven't dug into the blog post. I will check what's going on later. One thing I noticed is that, `emacs-version` is reported as `28.2.50` which makes the path `elpa/28.2.50/develop` but all the other packages spacemacs has installed are in `elpa/28.2/develop`.

**Comment by lebensterben** [comment]
make sure you removed all org-9.6* in your elpa directory.

**Comment by smile13241324** [comment]
Strange, this seems only to happen on emacs 28, not on emacs 29 and it seems to be independent from what is entered into the spacemacs buffer, setting it to show no content at all (no lists no agenda) has the same effect.

When I am trying to require org in my dotfile I am seeing the following:
```elisp
Debugger entered--Lisp error: (invalid-function org-assert-version)
  org-assert-version()
  byte-code("\300\301!\210\302 \210\300\303!\210\300\304!\207" [require org-macs org-assert-version cl-lib oc] 2)
  #<subr require>(org-keys)
  apply(#<subr require> org-keys)
  require(org-keys)
  eval-buffer(#<buffer  *load*> nil "/home/smile13241324/.emacs.d/elpa/28.2/develop/org..." nil t)  ; Reading at buffer position 3557
  load-with-code-conversion("/home/smile13241324/.emacs.d/elpa/28.2/develop/org..." "/home/smile13241324/.emacs.d/elpa/28.2/develop/org..." nil t)
  #<subr require>(org)
  apply(#<subr require> org)
  require(org)
  (progn (require 'org))
  eval((progn (require 'org)) t)
  elisp--eval-last-sexp(nil)
  #f(compiled-function (eval-last-sexp-arg-internal) "Evaluate sexp before point; print value in the echo area.\nInteractively, with a non `-' prefix argument, print output into\ncurrent buffer.\n\nThis commands handles `defvar', `defcustom' and `defface' the\nsame way that `eval-defun' does.  See the doc string of that\nfunction for details.\n\nNormally, this function truncates long output according to the\nvalue of the variables `eval-expression-print-length' and\n`eval-expression-print-level'.  With a prefix argument of zero,\nhowever, there is no such truncation.\nInteger values are printed in several formats (decimal, octal,\nand hexadecimal).  When the prefix argument is -1 or the value\ndoesn't exceed `eval-expression-print-maximum-character', an\ninteger value is also printed as a character of that codepoint.\n\nIf `eval-expression-debug-on-error' is non-nil, which is the default,\nthis command arranges for all errors to enter the debugger." (interactive "P") #<bytecode -0x45c973c4e971b99>)(nil)
  #f(compiled-function (&rest _it) #<bytecode 0x1fff02a74872>)()
  eval-sexp-fu-flash-doit-simple(#f(compiled-function (&rest _it) #<bytecode 0x1fff02a74872>) #f(compiled-function (&rest args2) #<bytecode -0x1133d5a2cfbd057f>) #f(compiled-function (&rest args2) #<bytecode -0x112bf59b0182057f>))
  eval-sexp-fu-flash-doit(#f(compiled-function (&rest _it) #<bytecode 0x1fff02a74872>) #f(compiled-function (&rest args2) #<bytecode -0x1133d5a2cfbd057f>) #f(compiled-function (&rest args2) #<bytecode -0x112bf59b0182057f>))
  esf-flash-doit(#f(compiled-function (&rest _it) #<bytecode 0x1fff02a74872>) #f(compiled-function (&rest args2) #<bytecode -0x1133d5a2cfbd057f>) #f(compiled-function (&rest args2) #<bytecode -0x112bf59b0182057f>) #f(compiled-function (&rest args2) #<bytecode 0x153e757edc0c5a87>))
  ad-Advice-eval-last-sexp(#f(compiled-function (eval-last-sexp-arg-internal) "Evaluate sexp before point; print value in the echo area.\nInteractively, with a non `-' prefix argument, print output into\ncurrent buffer.\n\nThis commands handles `defvar', `defcustom' and `defface' the\nsame way that `eval-defun' does.  See the doc string of that\nfunction for details.\n\nNormally, this function truncates long output according to the\nvalue of the variables `eval-expression-print-length' and\n`eval-expression-print-level'.  With a prefix argument of zero,\nhowever, there is no such truncation.\nInteger values are printed in several formats (decimal, octal,\nand hexadecimal).  When the prefix argument is -1 or the value\ndoesn't exceed `eval-expression-print-maximum-character', an\ninteger value is also printed as a character of that codepoint.\n\nIf `eval-expression-debug-on-error' is non-nil, which is the default,\nthis command arranges for all errors to enter the debugger." (interactive "P") #<bytecode -0x45c973c4e971b99>) nil)
  apply(ad-Advice-eval-last-sexp #f(compiled-function (eval-last-sexp-arg-internal) "Evaluate sexp before point; print value in the echo area.\nInteractively, with a non `-' prefix argument, print output into\ncurrent buffer.\n\nThis commands handles `defvar', `defcustom' and `defface' the\nsame way that `eval-defun' does.  See the doc string of that\nfunction for details.\n\nNormally, this function truncates long output according to the\nvalue of the variables `eval-expression-print-length' and\n`eval-expression-print-level'.  With a prefix argument of zero,\nhowever, there is no such truncation.\nInteger values are printed in several formats (decimal, octal,\nand hexadecimal).  When the prefix argument is -1 or the value\ndoesn't exceed `eval-expression-print-maximum-character', an\ninteger value is also printed as a character of that codepoint.\n\nIf `eval-expression-debug-on-error' is non-nil, which is the default,\nthis command arranges for all errors to enter the debugger." (interactive "P") #<bytecode -0x45c973c4e971b99>) nil)
  eval-last-sexp(nil)
  funcall-interactively(eval-last-sexp nil)
  call-interactively(eval-last-sexp nil nil)
  command-execute(eval-last-sexp)

```
I think we need more investigation into this first.

**Comment by nixmaniack** [comment]
I had to explicitly set the `package-user-dir` to a value that normal spacemacs sets up and I was able to install `org` with the instructions provided. For now, the workaround seems to work. Thanks @lebensterben!

```elisp
(require 'package)
(add-to-list 'package-archives '("melpa" . "https://melpa.org/packages/"))
(setq package-user-dir "/home/localuser/.emacs.d/elpa/28.2/develop/")
(package-refresh-contents)
(package-install 'org)
```

**Comment by hny-gd** [comment]
For me running

`cd ~/.emacs.d; rm -rf elpa/develop/org-9*` (note develop instead of 28.2 as @lebensterben wrote above)

and then starting Emacs was enough. 

Spacemacs then reinstalled org and everything was fine.

**Comment by emacs18** [comment]
There has been many problems like this over the years due to org mode.  Emacs comes with org mode, but almost always folks need to install a newer version.  This newer version is also needed by other packages as well.  That is why I make sure that org package is installed very early on in `early-init.el`.  I used to do this when I used `package.el` years ago.  Since I've been using `straight.el` for few years, I have `(straight-use-package 'org)` very early on in my `early-init.el` file to assure that org package is installed before even getting to most spacemacs initialization code.

**Comment by RidaAyed** [comment]
> I had to explicitly set the `package-user-dir` to a value that normal spacemacs sets up and I was able to install `org` with the instructions provided. For now, the workaround seems to work. Thanks @lebensterben!
> 
> ```emacs-lisp
> (require 'package)
> (add-to-list 'package-archives '("melpa" . "https://melpa.org/packages/"))
> (setq package-user-dir "~/.emacs.d/elpa/28.2/develop/")
> (package-refresh-contents)
> (package-install 'org)
> ```
Note that in my case, alongside the existing package directory (e.g. ~/.emacs.d/elpa/develop) the upper snippet **downloads and installs all packages** into the `package-user-dir` and finally reports: `‘org’ is already installed`

> make sure you removed all org-9.6* in your elpa directory.

`M-x org-version` still reports `Org mode version 9.5.5 (release_9.5.5 @ /usr/share/emacs/28.2/lisp/org/)`

**Comment by lebensterben** [comment]
@RidaAyed 

https://github.com/syl20bnr/spacemacs/issues/15896#issuecomment-1397716997

have you first removed org from elpa directory?

**Comment by emacs18** [comment]
I don't think org package is in melpa.  You need to add https://elpa.gnu.org/packages/.  Following is how spacemacs sets up packages.

```elisp
(defun configuration-layer/create-elpa-repository (name output-dir)
  "Create an ELPA repository containing all packages supported by Spacemacs."
  (configuration-layer/make-all-packages 'no-discover)
  (let (package-archive-contents
        (package-archives '(("melpa" . "https://melpa.org/packages/")
                            ("gnu"   . "https://elpa.gnu.org/packages/")
                            ("nongnu" . "https://elpa.nongnu.org/nongnu/"))))

```

**Comment by lebensterben** [comment]
if one followed my instructions melpa is added.

**Comment by notestaff** [comment]
FWIW, removing the byte-compiled *.elc files in the installed elpa org directory seems to fix the org-assert-function issue for me (but I use standard emacs not spacemacs).

**Comment by emacs18** [comment]
Corrupt byte compiled files could result if you installed org package after already having loaded built-in org.el.
This is why one must make sure that org.el is not loaded when you install org package!
This is why I have code in early-init.el (rather than init.el) to not only setup package archives, but also to install org packages.
Because this is done so early on during emacs start-up, there is no possibility of org.el being loaded when org package is installed.

I think this is the most common problem with org mode the folks have had over the years.

**Comment by lebensterben** [comment]
@emacs18 will there be any impact on loading time?

**Comment by emacs18** [comment]
I don't think so, but I am not sure.  Even if startup time is longer, I could tolerate it if that prevents hard to debug errors.

**Comment by tko** [comment]
I _think_ this might get fixed by deleting _all_ of `elpa/` -- no guarantees though. (Two computers, on one I deleted everything for unrelated reasons and it works ok, on the other I've just kept upgrading things and hit this same error. Was planning to delete elpa when I have the time.)

**Comment by sir4ur0n** [comment]
@tko I just tested and it doesn't solve the problem for me 😞

**Comment by RidaAyed** [comment]
> Inspired by https://irreal.org/blog/?p=10999
> 
> To fix it:
> 
>     1. `cd ~/.emacs.d; rm -rf elpa/28.2/develop/org-9*`
> 
>     2. start emacs with `emacs -Q` and evaluate (with `C-j`)
> 
> 
> ```emacs-lisp
> (require 'package)
> (add-to-list 'package-archives '("melpa" . "https://melpa.org/packages/"))
> (setq package-user-dir
>       (expand-file-name "develop"
>                         (expand-file-name emacs-version
>                                           (expand-file-name "elpa"
>                                                             user-emacs-directory))))
> (package-refresh-contents)
> (package-install 'org)
> ```
> 
>     3. Restart Emacs normally.

@lebensterben 

Running your snippet returns "‘org’ is already installed"

Yes, even deleted the complete elpa folder. Tried even deleting `/usr/share/emacs/28.2/lisp/org/` to no avail. After reinstalling emacs `yay emacs > reinstall` I'm back to "‘org’ is already installed"

@lebensterben  ..and of course: Thank you very much for your time :pray:

**Comment by lebensterben** [comment]
@RidaAyed 

what's your emacs version

**Comment by RidaAyed** [comment]
@lebensterben 

#### System Info :computer:
- OS: gnu/linux
- Emacs: 28.2
- Spacemacs: 0.999.0
- Spacemacs branch: develop (rev. 8645f21cb)
- Graphic display: t
- Running in daemon: nil
- Distribution: spacemacs
- Editing style: vim
- Completion: ivy

**Comment by emacs18** [comment]
@RidaAyed 

Can you explain how hour setup can install org package at all?

You set `package-archives` to include only melpa. As far as I can tell, melpa does not provide org package! Thus I don't see how you could possibly be installing org package when `package-archives` does not include https://elpa.gnu.org/packages/ as well.

**Comment by emacs18** [comment]
Sorry.  I just noticed that you are using add-to-list.  So you are probably relying on default value of package-archive to already have elpa.gnu.org.

**Comment by RidaAyed** [comment]
@emacs18 I was referring to to the [snippet](https://github.com/syl20bnr/spacemacs/issues/15896#issuecomment-1397716997) from @lebensterben 

Otherwise I'm relying on the standard spacemacs layers
```lisp
dotspacemacs-configuration-layers
'(
  (org :variables
       org-enable-hugo-support t
       org-enable-valign nil
       )
  )
```

**Comment by nixmaniack** [comment]
@RidaAyed Your snippet sets the value of the `package-user-dir` to a different value than what spacemacs computes. I had similar issues when I was debugging hence I hardcoded the value for the workaround. You can check how spacemacs computes the value [here](https://github.com/syl20bnr/spacemacs/blob/a529b0bfa8fb04c368e6d57059929806b9722e2f/core/core-configuration-layer.el#L440-L442) but in short it strips the patch version from the version string. So even though the workaround downloads the fixed org version, it's never used by Spacemacs. 

For clarity, on my machine, Spacemacs computes `package-user-dir` as `/home/localuser/.emacs.d/elpa/28.2/develop/`, and your snippet computes it as `/home/localuser/.emacs.d/elpa/28.2.50/develop`.

**Comment by lebensterben** [comment]
My snippet would work if one uses the default value of `dotspacemacs-elpa-subdirectory` and a stable release of Emacs.

otherwise, one should use https://github.com/syl20bnr/spacemacs/blob/a529b0bfa8fb04c368e6d57059929806b9722e2f/core/core-configuration-layer.el#L440-L442 to calculate the correct path and hardcode it as @nixmaniack pointed out.

**Comment by vitaminace33** [comment]
As of today, I did a clean install. After that I had to delete `org-9.6.1` and let Emacs reinstall it.

**Comment by RidaAyed** [comment]
I've changed `dotspacemacs-elpa-subdirectory` from `nil` to `emacs-version`

Describe variable `package-user-dir` returns:

      Its value is "/home/localuser/.emacs.d/elpa/28.2/develop/"
      Original value was "~/.emacs.d/elpa"

However deleting the complete elpa folder followed by the snippet from @lebensterben doesn't help. The snippet returns `org mode is already installed`

Since org-9.6 in fact resides in `/home/localuser/.emacs.d/elpa/28.2/develop/` after installation I can't understand why Spacemacs doesn't pick it up  and `org-version` remains 9.5.5.

@nixmaniack I couldn't reproduce how you got `/home/localuser/.emacs.d/elpa/28.2.50/develop`

**Comment by nixmaniack** [comment]
![image](https://user-images.githubusercontent.com/542810/222477140-746fefea-2f2c-4f30-979d-85469a0a5185.png)

So, using the first snippet to compute the value of `package-user-dir` makes the workaround useless because Spacemacs never looks in that directory (at least in my case since I'm using emacs version 28.2.50). 

I would suggest following:

1. Run Spacemacs normally, and check the value of `package-user-dir`.
2. Then, use the workaround snippet by hardcoding the value of the `package-user-dir`

**Comment by RidaAyed** [comment]
Thank you @nixmaniack

In summary:

1. Get `package-user-dir` in spacemacs. E.g. `"/home/localuser/.emacs.d/elpa/28.2/develop/"`
1. Hardcode returned value from step 1 in "workaround" snippet line 3
1. Delete `elpa` folder
1. Run "workaround" snippet e.g. `emacs --no-site-file --no-init-file --script fix-org-mode.el`

```lisp
;; fix-org-mode.el`
(require 'package)
(add-to-list 'package-archives '("melpa" . "https://melpa.org/packages/"))
(setq package-user-dir "/home/localuser/.emacs.d/elpa/28.2/develop/")
(package-refresh-contents)
(package-install 'org)
```

Yet, the problem remains: Org is already installed.

Interesting to note is that 9.6 is mentioned in brackets though in `M-x org-version`: `Org mode version 9.5.5 (release_9.5.5 @ /home/localuser/.emacs.d/elpa/28.2/develop/org-9.6.1/)`

**Comment by nixmaniack** [comment]
@RidaAyed That's strange. Do you have any org config that maybe getting triggered before Spacemacs initializes properly? You have to wrap such configs with `with-eval-after-load` which is a [documented issue](https://develop.spacemacs.org/layers/+emacs/org/README.html#important-note). I'm afraid I'm out of ideas after this.

**Comment by lebensterben** [comment]
@nixmaniack 

that should not matter because emacs -Q doesn't load any init script.

**Comment by nixmaniack** [comment]
@lebensterben My bad. You're right, emacs -Q won't load any Spacemacs' init code.

**Comment by RidaAyed** [comment]
Could the issue stem from `/usr/share/emacs/28.2/lisp/org/`?

In case someone has the same issue but knows lisp better than I do: https://github.com/jwiegley/use-package/issues/319

In particular I was interested in this snippet, though I'm getting a `void` error running this two lines using `emacs --no-site-file --no-init-file --script foo.el`:
```lisp
;;foo.el
(assq-delete-all 'org package--builtins)
(assq-delete-all 'org package--builtin-versions)
```

**Comment by emacs18** [comment]
I built emacs 28.3 just to look into this issue.  I see no problem at all.  Org 9.6.1 is installed the very first time I launched emacs 28.  I was able to load a dummy foo.org file without any error.  I can verify that org-mode.el that was loaded was the one installed by spacemacs rather than built-in.  I don't see why we need to mess with package--builtins or anything like it.

**Comment by emacs18** [comment]
Folks the packaging system deals with not only package names, but also package versions.  Spacemacs clearly states that it requires minimum of org version 9.6.1 via this line:
```elisp
    (org :location elpa :min-version "9.6.1")
```
This in turn is honored by functions such as `package-installed-p` so that the packaging system knows to installed 9.6.1, because the built-in version is older than this version.  At least this is the case for emacs 28 even if this is not the case for emacs 29 since in emacs 29 the built in version is either 9.6 or 9.6.1.

**Comment by emacs18** [comment]
The only potential problem is byte-compiling.  When newer org is installed and if built-in org mode has been loaded already, then the resulting byte-compiled files may be corrupt.  Easy fix for that is just remove all *.elc files in the installed org package directory and restart emacs which should re-compile all org elisp files properly.

**Comment by emacs18** [comment]
My testing used latest emacs 28.3 built on ubuntu 22.04, using latest spacemacs on develop branch, and using default set of layers except to add org layer.

**Comment by RidaAyed** [comment]
@emacs18 28.3? I got 28.2 installed as latest [release](https://www.gnu.org/software/emacs/history.html) on arch linux

**Comment by emacs18** [comment]
As I said before, I compiled my own emacs 28.3 from the latest source code.  I normally use my own build of emacs 29. 

I don't know whether emacs 28.3 was release or not yet.  The version number on the git branch is normally updated to be the forthcoming release, so I assume that 28.3 is not yet released.

**Comment by lebensterben** [comment]
it's 28.3 rc 1

**Comment by moritzschaefer** [comment]
Similarly to @RidaAyed I also got the 'emacs already installed issue'. However, just startiing spacemacs installed everything correctly

**Comment by RidaAyed** [comment]
Starting from scratch using `M-x dotspacemacs/install` was my remedy

**Comment by dr-hildi** [comment]
@RidaAyed: Can you please give some more details what stepps helped you in the end?

My setup is Debian 11 and Emacs 27.1.

1. `git clone https://github.com/syl20bnr/spacemacs ~/.emacs.d`
2. `rm .spacemacs` (thats what `M-x dotspacemacs/install` is doing, correct?)
3. `emacs` (installing 132 packages), Spacemacs up and running.
4. restored my `.spacemacs` file
5. restart installs another 157 packages

This again results in `(invalid-function org-assert-version)`. Calling `M-x dotspacemacs/install` deletes packages and Spacemacs is up again but without all my settings.

So this is just to bring the base installation back?

**Comment by emacs18** [comment]
I too see `invalid-function` problem using emacs 27.1.   However simply following the solution suggested multiple times before work, ie., simply remove byte-compiled files via 

```
rm ~/.emacs.d/elpa/27.1/develop/org*/*.elc
```
then restart emacs.  This is all that I needed to fix the problem on emacs 27.1.

Emacs 28 and 29 do not run into this problem as I have stated before.

**Comment by dr-hildi** [comment]
@emacs18 : I'm sorry. You are totally right. 

I tried this before without success. Trying to reproduce my problem last night it just worked as described with long grown and groomed `.spacemacs` config and emacs 27.1:

1. `git clone https://github.com/syl20bnr/spacemacs ~/.emacs.d`
2. `emacs`
3. `rm ~/.emacs.d/elpa/27.1/develop/org*/*.elc`
4. `emacs`

Done. Everything up and running.

**Comment by MHOOO** [comment]
> I too see `invalid-function` problem using emacs 27.1. However simply following the solution suggested multiple times before work, ie., simply remove byte-compiled files via
> 
> ```
> rm ~/.emacs.d/elpa/27.1/develop/org*/*.elc
> ```
> 
> then restart emacs. This is all that I needed to fix the problem on emacs 27.1.
> 
> Emacs 28 and 29 do not run into this problem as I have stated before.

Actually, I'm seeing this on emacs 28.3 (checked out yesterday via emacs-28 branch). Deleting the `*.elc` files worked to "fix" the issue after a restart, however.

**Comment by emacs18** [comment]
Did you see this problem with 28.3 when you followed the 5 step instruction detailed by @dr-hildi ?

**Comment by dr-hildi** [comment]
@emacs18 : I have to appologize. Spacemacs and Org-Mode are totally essential for me so I set up a 27.1 environment that works as mentioned above. I will set up a testenvironment for new 28.3 tests and will come back with my feedback if done. But this will take some time.

**Comment by emacs18** [comment]
@dr-hildi, my previous comment was addressed to @MHOOO.  What I'm curious about is whether @MHOOO  was able to see the problem on emacs 28 using your 5 steps or not.  If the problem was seen on emacs 28 by following some other steps, then perhaps we should try to track it down to understand it better.

**Comment by vatrat** [comment]
Same issue, OpenSUSE Tumbleweed 20230331 and emacs 28.2. I had resume-layouts on with org-mode buffers saved, which I fixed without getting any effect. I remembered my scratch-buffer was set to org-mode and unkillable. Once I changed scratch-buffer to text-mode it could start up most of the way. My path is ~/.emacs.d/elpa/28.2/develop/. I tried deleting all elpa packages with org in the name, then I ran emacs -Q and pasted the snippet in  lebensterben's first post, which had been updated by the time I copied it. This returned "‘org’ is already installed". Running M-x org-version in the same window returned 9.5.5, *however* once I closed it, starting spacemacs gave a message about org not being installed and org 9.6.2 was installed successfully. It worked as soon as the installation was done, but I restarted emacs anyway to be sure.

**Comment by tusharhero** [comment]
I still get the same error even after reinstalling spacemacs.  I also tried the workaround  by @lebensterben

**Comment by acllm** [comment]
I had to delete all my `org` related deps in `.spacemacs` to get rid of this error, since I rarely use `org` layer.
Here is what I did,
```shell
# comment `org` related deps
  ;; org
...
  ;; (org-tree-slide :location (recipe
  ;;                        :fetcher github
  ;;                        :repo "takaxp/org-tree-slide"))
  ;; org-mode
...
  ;; org-download
  ;; (setq-default org-download-heading-lvl nil)
  ;; (setq-default org-download-image-dir "./images")
  ;; (defun dummy-org-download-annotate-function (link)
  ;;   "")
  ;; (setq org-download-annotate-function
        ;; #'dummy-org-download-annotate-function)
  ;; org-tree-slide
  ;; (setq-default org-image-actual-width nil)
```

Then exit Emacs with `C-x C-c` and start.

**Comment by vatrat** [comment]
Well, I wish that had been the end of it. Today, I set `dotspacemacs-scratch-mode` back to `org-mode` (from `text-mode`). After reloading .spacemacs with `SPC f e R`, spacemacs failed to load (loading bar stuck at about 1/3 progress, but otherwise usable) with the error `byte-code: Invalid function: org-assert-version` seen from the `messages` buffer. I immediately switched back to `text-mode`, but no change. Restarting after that with `SPC q q` (`SPC q r` doesn't work with the way I have daemon set up), spacemacs is able to load fully but attempting to do something like `SPC b N n` gives the error `byte-code: Invalid function: org-assert-version`. I don't know why it would affect that, but restarting and updating packages doesn't help.
Addenda:
- Running `SPC f e c` to recompile packages produces pages and pages of "Invalid function: org-assert version" in the `compile-log` buffer from ob-\*, org-\*, and ox-\*.

UPDATE: It seems I accidentally set `dotspacemacs-new-empty-buffer-major-mode` from `text-mode` to `org-mode` instead of setting `dotspacemacs-scratch-mode`, so that would explain why new buffer wasn't working and the scratch buffer was. With my most recent package update in spacemacs, all org errors are fixed.

**Comment by bryce-carson** [comment]
I wasn't able to fix the issue using the accepted work-around. What I did to fix the issue was the following.

1. Download the latest version of Org from ELPA directly: [Org on ELPA](https://elpa.gnu.org/packages/org.html).
2. Open Emacs normally, that is, not in debug mode or anything else special. Just run `emacs` or click the desktop icon.
3. In a scratch buffer, run the following Emacs Lisp.
   ```lisp
   ;; Provide the filename of the version of Org mode which you downloaded.
   (call-interactively
    (lambda (f)
     "This lambda is part of a work-around for Spacemacs issue #15896."
      (interactive "fProvide path to tar file for Org mode:")
      (package-install-file f)))
   ```
4. <kbd>SPC</kbd> <kbd>f</kbd> <kbd>e</kbd> <kbd>c</kbd> to recompile packages.
5. Inspect the packages which produce warnings and errors and reinstall them with <kbd>Meta</kbd> <kbd>x</kbd> `package-reinstall`.
6. Repeat steps 4 and 5 until no warnings related to `org-` or `org--` prefixed functions or variables are present.
7. Restart Emacs.

It's definitely a strange issue.

Installing Org from ELPA using the package interface after setting `package-user-dir` correctly did not fix the issue, nor did recompiling at that stage.

The issue persisted until I installed from a TAR file, recompiled, reinstalled the other packages, and then finally restarted Emacs.

**Comment by sir4ur0n** [comment]
@bryce-carson there are 2 typos in your command :sweat_smile: 
```diff
- (package-install-from-file "~/Downloads/org.9.6.5.tar")
+ (package-install-file "~/Downloads/org-9.6.5.tar")
```

But otherwise thank you, **finally** the errors with Org are gone :pray:

**Comment by bryce-carson** [comment]
> @bryce-carson there are 2 typos in your command :sweat_smile: 
> ```diff
> - (package-install-from-file "~/Downloads/org.9.6.5.tar")
> + (package-install-file "~/Downloads/org-9.6.5.tar")
> ```
> 
> But otherwise thank you, **finally** the errors with Org are gone :pray: 

Ah, thanks! I corrected the typos. Glad it worked!

---

Update: the diff and these two comments are no longer relevant to the fix in my previous comment. See the previous comment from me, including the work-around, and ignore other discussion.

**Comment by maxnikulin** [comment]
> Corrupt byte compiled files could result if you installed org package after already having loaded built-in org.el.
> This is why one must make sure that org.el is not loaded when you install org package!

Actually any loaded Org version, not necessary built-in one, may cause mixed version load or even worse mixed version compilation issues. However emacs-29 branch contains changes in `package.el` that allow proper compiling of new Org version even when an older one has been loaded earlier. So at least the issue when manual removing of Org .elc files before recompiling should be alleviated.

(I am not a spacemacs user, I just spotted this issue in search engine results for an unrelated query.)

**Comment by Davidj361** [comment]
Still having this issue on the freshest pull of develop branch.

> @emacs18 : I'm sorry. You are totally right.
> 
> I tried this before without success. Trying to reproduce my problem last night it just worked as described with long grown and groomed `.spacemacs` config and emacs 27.1:
> 
>     1. `git clone https://github.com/syl20bnr/spacemacs ~/.emacs.d`
> 
>     2. `emacs`
> 
>     3. `rm ~/.emacs.d/elpa/27.1/develop/org*/*.elc`
> 
>     4. `emacs`
> 
> 
> Done. Everything up and running.

`rm ~/.emacs.d/elpa/27.2/develop/org*/*.elc` fixed my problem. Shouldn't this be fixed by now? I was getting: `(invalid-function org-assert-version)`

**Comment by vatrat** [comment]
Re-reporting this, now in a fresh Ubuntu 23.04 installation, emacs 28.2. Here is my [.spacemacs file](https://pastebin.com/z9ibMSRn). Initial spacemacs installation/compilation failed with invalid function: org-assert-version. Restarting spacemacs several times allowed it to completely start up, but still the org-assert-version error when intentionally trying to use org. I followed lebensterben's workaround, as I did in my last comment, and it again said "'org' is already installed", with "M-x org-version" from "emacs -Q" returning 9.5.5. Closing emacs and opening spacemacs, running "org-version" still gives an invalid function error. Running "SPC f e c" to recompile gives many "invalid function: org-assert-version" errors, about 30. Restarting spacemacs and updating packages did not change anything. Unclear on how to proceed.

**Comment by bryce-carson** [comment]
> Unclear on how to proceed.

@vatrat Please follow the steps I posted and report if the procedure works.

**Comment by vatrat** [comment]
@bryce-carson Thank you for the followup. After following your instructions completely, org-mode does work, reporting the correct version with M-x org-version. The regex in the list did not work, so I had to replace it with the path directly.
To repeat, this is a workaround, though I can't understand how it works

**Comment by maxnikulin** [comment]
> It's definitely a strange issue. Installing Org from ELPA using the package interface after setting `package-user-dir` correctly did not fix the issue, nor did recompiling at that stage. Until I installed from a TAR file, recompiled, reinstalled the other packages, and then finally restarted, the issue did not go away.

I do not think, installing directly from tar is relevant.
```
invalid function: org-assert-version
```
means that Org was compiled when an old (likely built-in) version of Org is already loaded (through dependency, when opening an Org file, because help autocompletion offered a function or variable from Org, etc.). `batch-byte-compile` and `byte-recompile-directory` functions can not help to recover from such state. Remove Org `.elc` files manually or delete org package.

The following commit from the emacs-29 branch fixes update of loaded package by `package.el`:
https://git.savannah.gnu.org/cgit/emacs.git/commit/?id=9dfd945a2c2
https://debbugs.gnu.org/cgi/bugreport.cgi?bug=49708
"Fix byte compilation of package built-ins"

A workaround for emacs < 29 may be unzipping `/usr/share/emacs/<VERSION>/lisp/org/*.el.gz` files. It should fix loading of `.el` files from the update over earlier loaded `.elc` files. If you do not like such trick, ensure that `(featurep 'org)` is evaluated to `nil` when your are going to install new Org version from ELPA, either from `list-packages` or from a file.

**Comment by Davidj361** [comment]
I can confirm that I got this problem with a version of emacs 27.2.

**Comment by bryce-carson** [comment]
> @bryce-carson The regex in the list did not work, so I had to replace it with the path directly.

Ah, yes. I'll edit that because it's not intended to be an actual regular expression. It's supposed to read as a user-supplied value (the path to the downloaded file, as you learned).

---

_All readers_: The current work-around, if you are having troubles following the discussion (because of edits or otherwise), is in this comment: https://github.com/syl20bnr/spacemacs/issues/15896#issuecomment-1530129761.

**Comment by bryce-carson** [comment]
> > It's definitely a strange issue. Installing Org from ELPA using the package interface after setting `package-user-dir` correctly did not fix the issue, nor did recompiling at that stage. Until I installed from a TAR file, recompiled, reinstalled the other packages, and then finally restarted, the issue did not go away.
> 
> I do not think, installing directly from tar is relevant.
> 
> ```
> invalid function: org-assert-version
> ```
> 
> means that Org was compiled when an old (likely built-in) version of Org is already loaded (through dependency, when opening an Org file, because help autocompletion offered a function or variable from Org, etc.). `batch-byte-compile` and `byte-recompile-directory` functions can not help to recover from such state. Remove Org `.elc` files manually or delete org package.
> 
> The following commit from the emacs-29 branch fixes update of loaded package by `package.el`: https://git.savannah.gnu.org/cgit/emacs.git/commit/?id=9dfd945a2c2 https://debbugs.gnu.org/cgi/bugreport.cgi?bug=49708 "Fix byte compilation of package built-ins"
> 
> A workaround for emacs < 29 may be unzipping `/usr/share/emacs/<VERSION>/lisp/org/*.el.gz` files. It should fix loading of `.el` files from the update over earlier loaded `.elc` files. If you do not like such trick, ensure that `(featurep 'org)` is evaluated to `nil` when your are going to install new Org version from ELPA, either from `list-packages` or from a file.

If that also works, that is good for a select few users. The current work-around, however, does not rely on super-user privileges for modifying system files; some distributions may detect changes in system files associated with packages and cause issues with system package managers. I'd prefer to avoid that.

Regardless, isn't the issue with the _byte-compiled files_ in that directory, not the compressed source files? If I knew more about how Emacs and package.el worked I could tinker with a fool-proof, portable solution, but as I mentioned previously (no longer visible due to comment edits) it's not worth the time investment. The current work-around is portable enough that if a user wants a work-around, if they've found their way here, they are more likely to have success running the simple commands until the errors disappear.

Keeping the work-around as simple as possible, even if it has more steps or takes another minute, is favourable because it introduces less headache. Reinstalling multiple packages, granted, may cause an issue in an exotic package (but given it's being reinstalled because of dependency issues there was kludge already), but in general it should be safe and portable.

**Comment by maxnikulin** [comment]
> > ```
> > invalid function: org-assert-version
> > ```
I forgot to mention that the "invalid function" means that `org-assert-version` is defined as a macro in current session, so new `org-macs.el` file is loaded, but during compilation `org-assert-version` was undefined because of old `org-macs.el` from built-in Org, so instead of macro expansion, a function call is added to the `.elc` file caused the error. Do not confuse it with
```
Symbol’s function definition is void: org-assert-version
```
that means an attempt to load a new incorrectly compiled `.elc` file while old `org-macs` is loaded. It means that for some reason built-in Org is loaded before the directory with new Org is added to `load-path'.

> Regardless, isn't the issue with the _byte-compiled files_ in that directory, not the compressed source files?

`.elc` files in `~/emacs.d/elpa/org-*/` are incorrectly compiled because of buggy `package.el` during installation of the ELPA package did not load updated `.el` files over built-in `/usr/share/emacs/27.1/lisp/org/*.elc` before compilation. The same `package.el` code is able to load new `.el` files when there are `.el` (not `.el.gz`) files in the same `/usr/share` directory and they are regular files, not symliks like in the elpa-org deb package.

> Keeping the work-around as simple as possible, even if it has more steps or takes another minute, is favourable because it introduces less headache.

I do not see a reason to require installing namely from **tar file**. Either there are more bugs in `package.el` (perhaps not fixed yet in the development branch) or it is not necessary.

**Comment by maxnikulin** [comment]
Does `spacemacs` have some specifics that makes the following workarounds unusable?

1. Install from a clean Emacs session
```
emacs -q
M-x package-delete RET org RET
M-x list-packages RET / n org RET
```
select Org, `i x` to install. Of course, the delete step is not required if the Org ELPA package has not been installed yet.

2. Alternatively remove Org `.elc` files and compile them (ignoring init file) from command line
```
rm -v ~/.emacs.d/elpa/org-[0-9]*/*.elc
emacs -q --batch -L ~/.emacs.d/elpa/org-[0-9]*/ -f batch-byte-compile ~/.emacs.d/elpa/org-[0-9]*/*.el
```
Do not try `byte-recompile-directory` instead, it may load `.elc` files incorrectly compiled by `package-install`.

**Comment by smile13241324** [comment]
Nope there is nothing hindering users to use this workaround except that the issue may reappear each time the org package is updated :(.

To be honest I do not see much we can do to work around this problem, in principle its a package.el bug which has been fixed there already, and will not appear in emacs 29.1 users should consider upgrading rather then doing all these workarounds in my humble opinion.

**Comment by bryce-carson** [comment]
> Nope there is nothing hindering users to use this workaround except that the issue may reappear each time the org package is updated :(.
> 
> To be honest I do not see much we can do to work around this problem, in principle its a package.el bug which has been fixed there already, and will not appear in emacs 29.1 users should consider upgrading rather then doing all these workarounds in my humble opinion.

Indeed. I'd say once Ubuntu and Fedora both ship 29.1 in the official repo, Spacemacs should no longer officially support ealier versions of Emacs.

Perhaps we should consider advising users install Guix and then they'd get access to 29.1 immediately, rather than waiting for other distros to ship different builds of the same upstream tag of Emacs, we'd rely on one, reproducible build?

**Comment by bryce-carson** [comment]
@smile13241324 For now, could you pin this issue, Maxi?, if you're able? It's relevant enough that it should have more visibility.

**Comment by smile13241324** [comment]
@bryce-carson, pinned. Yes indeed once 29.1 is officially released we should set the minimum version to 29.1 at least until there is an Ubuntu and Debian build available for it.

**Comment by vatrat** [comment]
Just chiming in, once/if we are able to find an automatic workaround for this issue, we should implement it. Otherwise, add a check in spacemacs to abort startup and show a message warning that the current version is less than 29.1 with a link to this issue. This way, it would prevent the bug from making spacemacs unusable until the user acknowledges the issue. I see this as a more useful fix than setting a minimum version, as it's going to be *quite a while* before 29.x is propagated to every distribution and every user updates. In the meantime, what happens when someone goes to open spacemacs after a while and it unexpectedly bricks itself? Dropping support for previous versions only introduces more support headaches, as people *will* continue to report this bug for as long as it occurs in the wild. We have access to elisp and version checks, and I think we should use them. This is an unfortunate situation that we shouldn't have to deal with, but let's make the most of it.

I guess that I'm just opposed to focusing on popular distributions because I started using spacemacs on a chromebook in a free VM and all other cursed manner of setups. If we want the latest version to be a hard requirement, we need to direct new users on how to install it on any platform, even if that's just dropping a link to emacs' installation guide.

**Comment by bryce-carson** [comment]
> Just chiming in, once/if we are able to find an automatic workaround for this issue, we should implement it. Otherwise, add a check in spacemacs to abort startup and show a message warning that the current version is less than 29.1 with a link to this issue. This way, it would prevent the bug from making spacemacs unusable until the user acknowledges the issue. I see this as a more useful fix than setting a minimum version, as it's going to be _quite a while_ before 29.x is propagated to every distribution and every user updates. In the meantime, what happens when someone goes to open spacemacs after a while and it unexpectedly bricks itself? Dropping support for previous versions only introduces more support headaches, as people _will_ continue to report this bug for as long as it occurs in the wild. We have access to elisp and version checks, and I think we should use them. This is an unfortunate situation that we shouldn't have to deal with, but let's make the most of it.
> 
> I guess that I'm just opposed to focusing on popular distributions because I started using spacemacs on a chromebook in a free VM and all other cursed manner of setups. If we want the latest version to be a hard requirement, we need to direct new users on how to install it on any platform, even if that's just dropping a link to emacs' installation guide.

The plan is to change the minimum supported version in Spacemacs itself, which after some minor refactoring, should refuse to initialize on ealier versions of Emacs. That's not a problem for any motivated user of an earlier Emacs version. Spacemacs is not currently bricked, and Spacemacs will not brick itself in any case. The problem is actually quite minor.

Closed issues can remain pinned, and documentation, guides, and READMEs serve better the purpose of communicating known issues to users.

The only real fix would be a labourious backport. The current work-around has a much lower development threshold; granted, a better and automatic solution would be preferred, but there are considerations to make.

First, should we prevent users from updating Org automatically? If so, we can hold the package by specifiying _a version_ string for it in `early-init.el`:

```emacs-lisp
(customize-set-variable 'package-load-list '(all (org "9.6.6")))
```

The consideration then becomes, "What version?" We might leave this up to users to decide, simply providing them the instruction so they only need to perform the current work-around procedure once, hold the package, and then forget about the issue.

An automated procedure is too delicate to spend our time on, I feel.

Another solution is to prevent Org from loading at all by setting the cdr of the list containing the symbol `org` to `nil`: `(customize-set-variable 'package-load-list '(all (org nil)))`. This will prevent _all_ versions of Org from being loaded at startup, and then Spacemacs _might_ be able to initialize a reinstallation automatically, and only once before holding the package, or simply loading a freshly compiled version of Org or *explicitly* loading Emacs Lisp files only for Org, and never byte compiling the package.

There's multiple solutions, but I'm not sure which would be best. I think instructing users to hold the Org package at a desired version using the above form after following the current work-around procedure is easiest.

**Comment by vatrat** [comment]
@bryce-carson I agree with you. I'm thinking that refusing to load org may be the best option, as it preserves other functionality by quarantine, disabling org rather than trying to make spacemacs fix the issue. This way, people who don't use org are unaffected and if there's still demand for a fix, pre-29.1 users can work to explore other options. In my eyes, the main benefit of this solution is making spacemacs "just work" for people who don't understand the inner workings.

**Comment by bryce-carson** [comment]
As a follow-up, rather than editing, when Spacemacs drops support for a version of Emacs it only means that if the reported issue is _caused_ by something in that version, such as the present issue, we probably won't make an effort to provide a work-around.

This issue is unique because it's caused by package.el and interacts exotically with Org. Other packages aren't suffering this issue (yet). In the current supported versions of Emacs, 27.1 or 28.2, the question "Is it an issue with Org, or package.el? Which do we address?" can be asked.

As I said in the previous comment, the fixes to package.el in 29.1 could be back-ported, which is labourious indeed. It's also fidddlesome, apart from the labour needed. It could introduce *more* issues.

It might be easier to hold Org to the 9.5 series, especially for new installations, and if a user wishes to upgrade Org then the work-around and further package version hand-holding would be advised.

**Comment by vatrat** [comment]
Hmm... So keeping org at 9.5.x by default (if a version below 29.1 is detected) would work? If so, I agree, that's a better idea. Is it possible to limit this behavior to only pre-29.1 versions, so that new installs on 29.1 or above don't get stuck with an old org package?

**Comment by bryce-carson** [comment]
It's trivial, actually. We'd simply add the following to `early-init.el`.

```emacs-lisp
(when (version< emacs-version "29.1")
  (customize-set-variable 'package-load-list '(all (org "9.5.5"))))
```

I'm not sure how related to the present issue it is, the upstream package website formerly published its own ELPA archive for Org packages (see https://orgmode.org/elpa.html). Org 9.5 was the last package version to work with this (Org 9.5.5 is distributed from GNU ELPA directly), and Org ELPA itself might become deprecated and offlined in the future. While we wait... a year? for 29.1 to land in Ubuntu (Fedora will have it in 39 if Emacs 29.1 gets released ahead of schedule) we might find other issues caused by Org ELPA's past.

> [Org 9.5 will be the last release to be distributed on Org ELPA.  After
9.5, we won't use Org ELPA for new releases.  Past releases will still
be available, though maybe not indefinitely.](https://list.orgmode.org/87blb3epey.fsf@gnu.org/)

With the "trivial" addition to early-init, we need to ask whether this will causes issues in sites with more recent versions of Org already installed, and how that will interact with byte compilation on 28.2, the portable dumper on 27.1, etc.

---

Edit: clarified a clause.
> Org 9.5 was the last package version to work with this (Org 9.5.5 is distributed from GNU ELPA directly)

**Comment by vatrat** [comment]
> With the "trivial" addition to early-init, we need to ask whether this will causes issues in sites with more recent versions of Org already installed, and how that will interact with byte compilation on 28.2, the portable dumper on 27.1, etc.

Oh boy, I'm starting to see what you mean. In the meantime, applying that "trivial" fix to new installations seems safe, assuming spacemacs has a way of detecting that.

**Comment by bryce-carson** [comment]
> > With the "trivial" addition to early-init, we need to ask whether this will causes issues in sites with more recent versions of Org already installed, and how that will interact with byte compilation on 28.2, the portable dumper on 27.1, etc.
> 
> Oh boy, I'm starting to see what you mean. In the meantime, applying that "trivial" fix to new installations seems safe, assuming spacemacs has a way of detecting that.

Indeed. That's why communication is easier. Spacemacs does not have as many active hackers as Doom Emacs does, and even fewer contributors understand it all.

The best way to ensure new installations include the change is to provide the code to users in the installation instructions. Feel free to submit a PR against the README if you want to get involved. Spacemacs needs more active contributors hacking on and proofing the documentation.

**Comment by vatrat** [comment]
Appreciate the response at all, I'll look at it when I'm done with my current work.

**Comment by bryce-carson** [comment]
> Appreciate the response at all, I'll look at it when I'm done with my current work.

Sounds good. I'm investigating the consequences of shipping the change by default (27.1 and such).

**Comment by maxnikulin** [comment]
> In the current supported versions of Emacs, 27.1 or 28.2, the question "Is it an issue with Org, or package.el?

The primary issue with Org is that it is a large package with multiple files and **macros** are actively used. The latter is a challenge for Emacs byte compilation. Void/invalid function issues were not rare reports on the Org mailing list, and it is a reason for the https://orgmode.org/worg/org-faq.html#mixed-install note. The idea of `org-assert-version` was to make clear to users that something is wrong in their configuration. I do not think that at the moment when it was introduced, it was clear that `package.el` can not handle update of a loaded package.

> So keeping org at 9.5.x by default (if a version below 29.1 is detected)

Wouldn't it better to keep the built-in version of Org?

**Comment by maxnikulin** [comment]
> once/if we are able to find an automatic workaround for this issue, we should implement it.

Once `invalid function: org-assert-version` is detected, temporary suppress loading of `.elc` files (`load-suffixes`), reload `org-macs.el`,  run `byte-recompile-directory` for Org.

**Comment by bryce-carson** [comment]
> > In the current supported versions of Emacs, 27.1 or 28.2, the question "Is it an issue with Org, or package.el?
> 
> The primary issue with Org is that it is a large package with multiple files and **macros** are actively used. The latter is a challenge for Emacs byte compilation. Void/invalid function issues were not rare reports on the Org mailing list, and it is a reason for the https://orgmode.org/worg/org-faq.html#mixed-install note. The idea of `org-assert-version` was to make clear to users that something is wrong in their configuration. I do not think that at the moment when it was introduced, it was clear that `package.el` can not handle update of a loaded package.
> 
> > So keeping org at 9.5.x by default (if a version below 29.1 is detected)
> 
> Wouldn't it better to keep the built-in version of Org?

That would be the purpose behind holding the package, to prefer the built-in package. On most platforms, that seems to be 9.5.5; for any platform that has a version less than 9.5, it's okay to uprade to 9.5.x, just not to 9.6.

For systems where 9.5.5 is distributed, it keeps the built-in, otherwise we allow upgrade to the newer version that doesn't produce the error.

That is only if we introduce this breaking change, however.

**Comment by wztdream** [comment]
I encountered the problem while upgrading Spacemacs and was able to resolve it by deleting the old org files. However, I encountered some errors initially, which were resolved after restarting Spacemacs. Although I accidentally resolved the problem, I am concerned that it may resurface later. Therefore, I believe it is necessary to investigate the issue related with the old org files in more detail.

**Comment by allentiak** [comment]
I've just bumped into this (ironically) after applying " Update htmlize \ Fix for the org version conflict." https://github.com/syl20bnr/spacemacs/commit/6e5816b82fdaaf7e887ad6b12b53eef7f666d810

Rolling back my packages to their 2023.07.12 version solved the problem.
That's why I always keep at least the last two updates in my "able-to-rollback" cache :)

**Comment by moritzschaefer** [comment]
I tried @lebensterben and other ways to fix this but no luck. org-version returns the new correct version 9.6.7 but still I get (invalid-function org-assert-version)

**Comment by 3dimaging** [comment]
> @lebensterben I followed your workaround but looks it didn't work for me. I did `rm -rf` the org packages but the last line to install `org` throws this message.
> 
> ```
> "‘org’ is already installed"
> ```
> 
> I just followed your exact instruction and haven't dug into the blog post. I will check what's going on later. One thing I noticed is that, `emacs-version` is reported as `28.2.50` which makes the path `elpa/28.2.50/develop` but all the other packages spacemacs has installed are in `elpa/28.2/develop`.

It does not matter. I followed @lebensterben and also got the message, "‘org’ is already installed". I restart the emacs. The org model works fine.

**Comment by neverkas** [comment]
I did what said @lebensterben and worked until I restart spacemacs so I decided to install Emacs 29 because of someone here said that there was not problem and now it works.

**Comment by biocyberman** [comment]
In addition to installing Emacs 29.1, I had to put the following into `(defun dotspacemacs/user-init ()` in `.spacemacs`
```
  (straight-use-package '(org :type built-in))
```

**Comment by felixwiemuth** [comment]
**Workaround in emacs 29.3**: After nothing else worked (I tried many of the suggestions here, but I am not using `straight`), my solution was to install the org package in `emacs -q` and then move the directory `.emacs.d/elpa/org-9.7.4` into `.emacs.d/elpa/29.3/` (also the `org-9.7.4.signed` file). Then this version is loaded by spacemacs.

Potentially I first removed org with `package-list-packages` or `package-delete` but that didn't seem to work. The package `org` was simply shown as available with the (then) current version `9.7.4` in `package-list-packages`.

**Problems:** I discovered some issues with cycling (it would sometimes close outer trees) and inserting new subtrees (sometimes inserting at random other places) but I am not sure whether this is due to the new version (and interaction with other packages) or both org versions being loaded as others mentioned in this thread (I am not loading anything org-specific in `.spacemacs`, only declaring the layer).
- Adding `(org :location elpa :min-version "9.7.5")` to `early-init.el` results in a warning ("Symbol's function definition is void: org"), so doesn't seem to have any effect
- Adding `(customize-set-variable 'package-load-list '(all (org "9.7.5")))` to `early-init.el` doesn't seem to have any effect either

**Regarding updating:** When there is an update for the `org` package (here `9.7.5`), spacemacs will suggest to update it but will instead remove the manually added directory without installing the new. Then the emacs-shipped org version will be loaded again. So one has to repeat the manual install process as described above.

Are there any other experiences with this issue for emacs 29.3?

**Comment by atparkweb** [comment]
> FWIW, removing the byte-compiled *.elc files in the installed elpa org directory seems to fix the org-assert-function issue for me (but I use standard emacs not spacemacs).

That solved the issue for me (Spacemacs 0.999.0@28.2). Thanks!
